### PR TITLE
Make LockServiceImpl.close idempotent and enable assertions on tests

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: The following stacktrace was thrown in an internal testing build:
```
ERROR [2017-07-12 11:31:34,814] [Test worker] com.palantir.util.JMXUtils - Assertion Failed to unregister mbean for name com.palantir.lock:type=LockServer_52 with exception javax.management.InstanceNotFoundException: com.palantir.lock:type=LockServer_52
at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.getMBean(DefaultMBeanServerInterceptor.java:1095)
at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.exclusiveUnregisterMBean(DefaultMBeanServerInterceptor.java:427)
at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.unregisterMBean(DefaultMBeanServerInterceptor.java:415)
at com.sun.jmx.mbeanserver.JmxMBeanServer.unregisterMBean(JmxMBeanServer.java:546)
at com.palantir.util.JMXUtils.unregisterMBeanCatchAndLogExceptions(JMXUtils.java:155)
at com.palantir.lock.impl.LockServiceImpl$4.run(LockServiceImpl.java:246)
at com.palantir.lock.impl.LockServiceImpl.close(LockServiceImpl.java:1067)
...
```

We believe that making LockServiceImpl.close idempotent should fix this problem. We're also enabling assertions on our tests, since this was the reason for our test infrastructure not catching this.

**Implementation Description (bullets)**: -

**Concerns (what feedback would you like?)**: Should I make `isShutDown` an atomic boolean?

**Where should we start reviewing?**: -

**Priority (whenever / two weeks / yesterday)**: -

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2144)

### CODE REVIEW

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java [hunk 1]
```diff
@@ -1061,10 +1061,12 @@ private StringBuilder getGeneralLockStats() {
...
+        if (!isShutDown) {
+            executor.shutdownNow();
+            wakeIndefiniteBlockers();
+            callOnClose.run();
+        }
         isShutDown = true;
-        executor.shutdownNow();
-        wakeIndefiniteBlockers();
-        callOnClose.run();
     }
```
Comment (jeremyk-91) [inline lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: Hmm. This generally fixes the issue, but is not thread-safe; if two threads call `close()` concurrently both could enter the if-block, and `callOnClose` is an arbitrary Runnable so not necessarily idempotent.

Minimal fix could be to synchronize this (people shouldn't depend on close being super performant). A better fix could actually be the AtomicBoolean idea you mention!

File: gradle/shared.gradle [hunk 2]
```diff
@@ -90,6 +90,8 @@ tasks.withType(Test) {
         reports.junitXml.destination = new File(System.env.CIRCLE_TEST_REPORTS, it.getName())
     }
 
+    enableAssertions = true
```
Comment (jeremyk-91) [inline gradle/shared.gradle]: Might be worth considering the perf impact of this on the builds

Comment (jeremyk-91) [inline gradle/shared.gradle]: (concrete build times seem unaffected, so we're ok here)

Comment (hsaraogi) [inline gradle/shared.gradle]: Yup, assertions dont really add any additional cost except the evaluation of boolean expressions. Build times should be a major concern here.

**Comment by jeremyk-91** [comment]
@hsaraogi yep, that's right. Currently it works because if I'm not wrong `close()` is never called on the TMs that were created, which is pretty hacky

**Pull request merged**